torchrl/envs/model_based.py

```diff
@@ -0,0 +1,163 @@
+        - input_spec (CompositeSpec): sampling spec of the actions and/or other inputs;
+        - reward_spec (TensorSpec): sampling spec of the rewards;
+        - batch_size (torch.Size): number of environments contained in the instance;
+        - device (torch.device): device where the env input and output are expected to live
+        - is_done (torch.Tensor): boolean value(s) indicating if the environment has reached a done state since the
+            last reset
+
+    Args:
+        world_model (nn.Module): model that generates world states and its corresponding rewards;
+        device (torch.device): device where the env input and output are expected to live
+        dtype (torch.dtype): dtype of the env input and output
+        batch_size (torch.Size): number of environments contained in the instance
+
+    Methods:
+        step (TensorDict -> TensorDict): step in the environment
+        reset (TensorDict, optional -> TensorDict): reset the environment
+        set_seed (int -> int): sets the seed of the environment
+        rand_step (TensorDict, optional -> TensorDict): random step given the action spec
+        rollout (Callable, ... -> TensorDict): executes a rollout in the environment with the given policy (or random
+            steps if no policy is provided)
+
+    """
+
+    def __init__(
+        self,
+        world_model: Union[nn.Module, List[TensorDictModule], TensorDictSequence],
+        device: DEVICE_TYPING = "cpu",
+        dtype: Optional[Union[torch.dtype, np.dtype]] = None,
+        batch_size: Optional[torch.Size] = None,
+    ):
+        super(ModelBasedEnv, self).__init__(
+            device=device, dtype=dtype, batch_size=batch_size
+        )
+        self.world_model = world_model
+
+    def set_specs_from_env(self, env: EnvBase):
+        """
+        Sets the specs of the environment from the specs of the given environment.
+        """
+        self.observation_spec = env.observation_spec
+        self.action_spec = env.action_spec
+        self.reward_spec = env.reward_spec
+
+    def _step(
+        self,
+        tensordict: TensorDict,
+    ) -> TensorDict:
+        # step method requires to be immutable
+        tensordict_in = tensordict.clone()
+        # Compute world state
+        tensordict_out = self.world_model(tensordict_in)
+        # Step requires a done flag. No sense for MBRL so we set it to False
+        tensordict_out["done"] = torch.zeros(tensordict_out.shape, dtype=torch.bool)
+        return tensordict_out
+
+    @abc.abstractmethod
+    def _reset(self, tensordict: TensorDict, **kwargs) -> TensorDict:
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def _set_seed(self, seed: Optional[int]) -> int:
+        raise NotImplementedError
+
+    def to(self, device: DEVICE_TYPING) -> ModelBasedEnv:
+        super().to(device)
+        self.world_model.to(device)
+        return self
+
+
+class DreamerEnv(ModelBasedEnv):
```

Comment:
Not sure if this belongs here or in examples... let's think about it

Reply:
FWIW, I think it seems more appropriate in an examples folder.